protocol/contracts/v2/circuitBreaker/TempleCircuitBreakerAllUsersPerPeriod.sol

```diff
@@ -0,0 +1,169 @@
+pragma solidity ^0.8.17;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Temple (v2/circuitBreaker/TempleCircuitBreakerAllUsersPerPeriod.sol)
+
+import { CommonEventsAndErrors } from "contracts/common/CommonEventsAndErrors.sol";
+import { TempleElevatedAccess } from "contracts/v2/access/TempleElevatedAccess.sol";
+import { SafeCast } from "contracts/common/SafeCast.sol";
+import { ITempleCircuitBreaker } from "contracts/interfaces/v2/circuitBreaker/ITempleCircuitBreaker.sol";
+
+/**
+ * @title Temple Circuit Breaker -- total volumes (across all users) in a rolling period window
+ * 
+ * @notice No more than the cap can be borrowed within a `periodDuration` window. 
+ * A slight nuance is that it will be slightly less than `periodDuration`, it also takes into account
+ * how many internal buckets are used - this is a tradeoff for gas efficiency.
+ * 
+ * -- The tracking is split up into hourly buckets, so for a 24 hour window, we define 24 hourly buckets.
+ * -- When a new transaction is checked, it will roll forward by the required buckets (when it gets to 23 it will circle back from 0), cleaning up the buckets which are now > 24hrs in the past.
+        If it's in the same hr as last time, then nothing to clean up
+ * -- Then adds the new volume into the bucket
+ * -- Then sums the buckets up and checks vs the cap, reverting if over.
+
+ * This means that we only have to sum up 24 items.
+
+ * The compromise is that the window we check for is going to be somewhere between 23hrs and 24hrs.
+ */
+contract TempleCircuitBreakerAllUsersPerPeriod is ITempleCircuitBreaker, TempleElevatedAccess {
+    using SafeCast for uint256;
+
+    /**
+     * @notice The duration of the rolling period window
+     */
+    uint32 public periodDuration;
+
+    /**
+     * @notice The maximum allowed amount to be transacted within each period
+     */
+    uint128 public cap;
+
+    /**
+     * @notice How many buckets to split the periodDuration into. 
+     * @dev A lower number of buckets means less gas will be used, however the rolling
+     * window will change at the *start* of the bucket time.
+     *   eg for a 24 hour `periodDuration`, with 24 (hourly) buckets, there could be a transaction
+     *   for the cap at the 13:45:00, and again the next day 24 hours later at 13:05:00
+     *   and this would be allowed.
+     * A higher number of buckets means this wait time is less, however this will use more gas.
+     */
+    uint32 public nBuckets;
+
+    /**
+     * @notice The derived length of time in each bucket
+     */
+    uint32 public secondsPerBucket;
+
+    /**
+     * @notice The current bucket index.
+     * @dev The first bucket starts at 1-1-1970
+     */
+    uint32 public bucketIndex;
+
+    /**
+     * @notice The total amount of volume tracked within each bucket
+     */
+    uint256[65535] public buckets;
```

Comment:
super nit: can you add a dev comment saying if `nBuckets` is beyong max value for `uint16`, `buckets` can't contain

Reply:
Fixed in https://github.com/TempleDAO/temple/pull/814/commits/3933dcacf45c04e2bb680b56d4921a0d6b6ba65a